### PR TITLE
Functions2

### PR DESCRIPTION
**Pull request by zgordan-vv**
Addresses part of #337

This change is [Reviewable](https://reviewable.io/reviews/unidoc/unioffice/348)

### CODE REVIEW

File: internal/wildcard/match.go [hunk 1]
```diff
@@ -81,3 +81,59 @@ func deepMatchRune(str, pattern []rune, simple bool) bool {
 	}
 	return len(str) == 0 && len(pattern) == 0
 }
+
+// Match -  finds whether the text matches/satisfies the pattern string.
+// supports  '*' and '?' wildcards in the pattern string.
+// unlike path.Match(), considers a path as a flat name space while matching the pattern.
+// The difference is illustrated in the example here https://play.golang.org/p/Ega9qgD4Qz .
+func Index(pattern, name string) (index int) {
```
Comment (gunnsth) [inline internal/wildcard/match.go]: Should be Index... ?

File: internal/wildcard/match.go [hunk 2]
```diff
@@ -81,3 +81,59 @@ func deepMatchRune(str, pattern []rune, simple bool) bool {
...
+		rname = append(rname, r)
+	}
+	for _, r := range pattern {
+		rpattern = append(rpattern, r)
+	}
+	simple := false // Does extended wildcard '*' and '?' match.
+	return deepIndexRune(rname, rpattern, simple, 0)
+}
+
+func deepIndexRune(str, pattern []rune, simple bool, counter int) int {
```
Comment (gunnsth) [inline internal/wildcard/match.go]: Add a short comment on what this does... Looking for pattern inside a string? why is deep needed/what does it mean?  What kind of index is returned, what kind of pattern is used?

File: spreadsheet/context.go [hunk 3]
```diff
@@ -116,6 +117,16 @@ func (e *evalContext) GetLocked(cellRef string) bool {
 	return e.s.Cell(cellRef).getLocked()
 }
 
+// HasFormula returns if the cell contains formula.
```
Comment (gunnsth) [inline spreadsheet/context.go]: returns true if ...

File: spreadsheet/context.go [hunk 4]
```diff
@@ -116,6 +117,16 @@ func (e *evalContext) GetLocked(cellRef string) bool {
...
+func (e *evalContext) HasFormula(cellRef string) bool {
+	return e.s.Cell(cellRef).HasFormula()
+}
+
+// IsBool returns if the cell boolean value.
```
Comment (gunnsth) [inline spreadsheet/context.go]: returns true if the cell is a boolean value

File: spreadsheet/context.go [hunk 5]
```diff
@@ -131,3 +142,33 @@ func (e *evalContext) GetWidth(colIdx int) float64 {
 	}
 	return 0
 }
+
+// GetEpoch returns a workbook's time epoch.
+func (e *evalContext) GetEpoch() time.Time {
+	return e.s.w.Epoch()
+}
+
+var dbcs []string = []string{
```
Comment (gunnsth) [inline spreadsheet/context.go]: what does dbcs stand for?

File: spreadsheet/formula/evaluator.go [hunk 6]
```diff
@@ -21,12 +26,62 @@ func NewEvaluator() Evaluator {
...
 func (d *defEval) Eval(ctx Context, formula string) Result {
 	expr := ParseString(formula)
 	if expr != nil {
+		d.addInfo(ctx, expr)
 		return expr.Eval(ctx, d)
 	}
 	return MakeErrorResult(fmt.Sprintf("unable to parse formula %s", formula))
 }
+
+func (d *defEval) addInfo(ctx Context, expr Expression) {
```
Comment (gunnsth) [inline spreadsheet/formula/evaluator.go]: Please add a short comment why this addInfo is needed. What kind of info is being added and why?
Seems like a pre-requisite for formula evaluation?

File: spreadsheet/formula/fninformation.go [hunk 7]
```diff
@@ -182,3 +197,160 @@ func Cell(ctx Context, ev Evaluator, args []Result) Result {
...
+
+	return MakeBoolResult(value != value/2*2)
+}
+
+// ISREF is an implementation of the Excel ISREF() function.
+func IsRef(ctx Context, ev Evaluator, args []Result) Result {
+	if len(args) != 1 {
+		MakeErrorResult("ISREF() accepts a single argument")
+	}
+//	refs := ev.(*defEval).args
```
Comment (gunnsth) [inline spreadsheet/formula/fninformation.go]: Remove commented code

File: internal/wildcard/match.go (hunk 1 above)
Comment (zgordan-vv) [reply internal/wildcard/match.go]: fixed

File: internal/wildcard/match.go (hunk 2 above)
Comment (zgordan-vv) [reply internal/wildcard/match.go]: fixed

File: spreadsheet/context.go (hunk 3 above)
Comment (zgordan-vv) [reply spreadsheet/context.go]: fixed

File: spreadsheet/context.go (hunk 4 above)
Comment (zgordan-vv) [reply spreadsheet/context.go]: fixed

File: spreadsheet/context.go (hunk 5 above)
Comment (zgordan-vv) [reply spreadsheet/context.go]: added description

File: spreadsheet/formula/evaluator.go (hunk 6 above)
Comment (zgordan-vv) [reply spreadsheet/formula/evaluator.go]: added

File: spreadsheet/formula/fninformation.go (hunk 7 above)
Comment (zgordan-vv) [reply spreadsheet/formula/fninformation.go]: removed

File: internal/wildcard/match.go [hunk 8]
```diff
@@ -81,3 +81,58 @@ func deepMatchRune(str, pattern []rune, simple bool) bool {
...
+	for _, r := range name {
+		rname = append(rname, r)
+	}
+	for _, r := range pattern {
+		rpattern = append(rpattern, r)
+	}
+	return deepIndexRune(rname, rpattern, 0)
+}
+
+//deepIndexRune is recursive function which searches for a position of substring which matches given wildcard pattern
```
Comment (gunnsth) [inline internal/wildcard/match.go]: Missing the space between / and deepIndexRune.
Also missing dot at the end of the sentence.
Also should write "is a recursive function"...

File: spreadsheet/context.go (hunk 5 above)
Comment (gunnsth) [reply spreadsheet/context.go]: Please use standard convention in comments.  Space between / and text. And sentences should always end with a dot.

File: spreadsheet/formula/evaluator.go (hunk 6 above)
Comment (gunnsth) [reply spreadsheet/formula/evaluator.go]: comment style

File: spreadsheet/formula/fndatetime.go [hunk 9]
```diff
@@ -0,0 +1,303 @@
+// Copyright 2017 FoxyUtils ehf. All rights reserved.
+//
+// Use of this source code is governed by the terms of the Affero GNU General
+// Public License version 3.0 as published by the Free Software Foundation and
+// appearing in the file LICENSE included in the packaging of this file. A
+// commercial license can be purchased by contacting sales@baliance.com.
```
Comment (gunnsth) [inline spreadsheet/formula/fndatetime.go]: Please ensure that this email address is NOT used anywhere.  We are sales@foxyutils.com

Comment (ahall) [reply spreadsheet/formula/fndatetime.go]: support@unidoc.io ?

Comment (zgordan-vv) [reply spreadsheet/formula/fndatetime.go]: replaced for support@unidoc.io

File: internal/wildcard/match.go (hunk 8 above)
Comment (zgordan-vv) [reply internal/wildcard/match.go]: fixed

File: spreadsheet/formula/evaluator.go (hunk 6 above)
Comment (zgordan-vv) [reply spreadsheet/formula/evaluator.go]: fixed

File: spreadsheet/context.go (hunk 5 above)
Comment (zgordan-vv) [reply spreadsheet/context.go]: fixed

**Pull request merged**